<!DOCTYPE html> <html lang="en-US"> <head> <meta charset="UTF-8"> <meta http-equiv="X-UA-Compatible" content="IE=Edge"> <title>Tutorial - ScreenBEAM</title> <link rel="shortcut icon" href="http://localhost:4000/favicon.ico" type="image/x-icon"> <link rel="stylesheet" href="http://localhost:4000/assets/css/just-the-docs.css"> <script type="text/javascript" src="http://localhost:4000/assets/js/vendor/lunr.min.js"></script> <script type="text/javascript" src="http://localhost:4000/assets/js/just-the-docs.js"></script> <meta name="viewport" content="width=device-width, initial-scale=1"> <!-- Begin Jekyll SEO tag v2.6.1 --> <title>Tutorial | ScreenBEAM</title> <meta name="generator" content="Jekyll v3.8.6" /> <meta property="og:title" content="Tutorial" /> <meta property="og:locale" content="en_US" /> <link rel="canonical" href="http://localhost:4000/docs/Tutorial.html" /> <meta property="og:url" content="http://localhost:4000/docs/Tutorial.html" /> <meta property="og:site_name" content="ScreenBEAM" /> <script type="application/ld+json"> {"url":"http://localhost:4000/docs/Tutorial.html","headline":"Tutorial","@type":"WebPage","@context":"https://schema.org"}</script> <!-- End Jekyll SEO tag --> </head> <body> <svg xmlns="http://www.w3.org/2000/svg" style="display: none;"> <symbol id="link" viewBox="0 0 16 16"> <title>Link</title> <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path> </symbol> </svg> <div class="page-wrap"> <div class="side-bar"> <div class="site-header"> <a href="http://localhost:4000" class="site-title lh-tight"> ScreenBEAM </a> <button class="menu-button fs-3 js-main-nav-trigger" data-text-toggle="Hide" type="button">Menu</button> </div> <div class="navigation main-nav js-main-nav"> <nav role="navigation" aria-label="Main navigation"> <ul class="navigation-list"><li class="navigation-list-item"><a href="http://localhost:4000/" class="navigation-list-link">Home</a></li><li class="navigation-list-item"><a href="http://localhost:4000/docs/Installation.html" class="navigation-list-link">Installation</a></li><li class="navigation-list-item active"><a href="http://localhost:4000/docs/Tutorial.html" class="navigation-list-link active">Tutorial</a></li><li class="navigation-list-item"><a href="http://localhost:4000/docs/Acknowledgements.html" class="navigation-list-link">Acknowledgements</a></li></ul> </nav> </div> <footer class="site-footer"> <p class="text-small text-grey-dk-000 mb-4">This site uses <a href="https://github.com/pmarsceill/just-the-docs">Just the Docs</a>, a documentation theme for Jekyll.</p> </footer> </div> <div class="main-content-wrap js-main-content" tabindex="0"> <div class="main-content"> <div class="page-header js-page-header"> <div class="search"> <div class="search-input-wrap"> <input type="text" class="js-search-input search-input" tabindex="0" placeholder="Search ScreenBEAM" aria-label="Search ScreenBEAM" autocomplete="off"> <svg width="14" height="14" viewBox="0 0 28 28" xmlns="http://www.w3.org/2000/svg" class="search-icon"><title>Search</title><g fill-rule="nonzero"><path d="M17.332 20.735c-5.537 0-10-4.6-10-10.247 0-5.646 4.463-10.247 10-10.247 5.536 0 10 4.601 10 10.247s-4.464 10.247-10 10.247zm0-4c3.3 0 6-2.783 6-6.247 0-3.463-2.7-6.247-6-6.247s-6 2.784-6 6.247c0 3.464 2.7 6.247 6 6.247z"/><path d="M11.672 13.791L.192 25.271 3.02 28.1 14.5 16.62z"/></g></svg> </div> <div class="js-search-results search-results-wrap"></div> </div> </div> <div class="page"> <div id="main-content" class="page-content" role="main"> <h1 class="no_toc" id="tutorial"> <a href="#tutorial" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Tutorial </h1> <p>ScreenBEAM2 is a R based tool which consists of three major parts for processing steps: 1. mapping long read sequence to short read libraries, 2. Quality control, data cleanning and data preprocessing for mapped raw counts data; 3. Differential representative analysis on gene level or shRNA level.</p> <h2 class="no_toc text-delta" id="table-of-contents"> <a href="#table-of-contents" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Table of contents </h2> <ol id="markdown-toc"> <li><a href="#step1mapping-to-library" id="markdown-toc-step1mapping-to-library">Step1.Mapping to library</a> <ol> <li><a href="#step11-prepare-raw-fastq-and-library-files" id="markdown-toc-step11-prepare-raw-fastq-and-library-files">Step1.1 Prepare raw fastq and library files</a></li> <li><a href="#step12-mapping-and-collecting-raw-counts" id="markdown-toc-step12-mapping-and-collecting-raw-counts">Step1.2 Mapping and collecting raw counts</a></li> <li><a href="#step13-mapping-qualtiy-control" id="markdown-toc-step13-mapping-qualtiy-control">Step1.3 Mapping qualtiy control</a></li> </ol> </li> <li><a href="#step2-data-annotation-and-data-cleanning" id="markdown-toc-step2-data-annotation-and-data-cleanning">Step2 Data annotation and data cleanning</a> <ol> <li><a href="#step21-create-metadata" id="markdown-toc-step21-create-metadata">Step2.1 Create Metadata</a></li> <li><a href="#step22-combine-mapping-info-to-your-metadata" id="markdown-toc-step22-combine-mapping-info-to-your-metadata">Step2.2 Combine mapping info to your metadata</a></li> <li><a href="#step23-data-normalization-and-create-expressionset" id="markdown-toc-step23-data-normalization-and-create-expressionset">Step2.3 Data normalization and create Expressionset</a></li> <li><a href="#step24-qualtiy-control-for-wrapped-expressionsets" id="markdown-toc-step24-qualtiy-control-for-wrapped-expressionsets">Step2.4 Qualtiy control for wrapped expressionsets</a></li> <li><a href="#step25-prepare-tsv-file-for-differential-representation-analysis" id="markdown-toc-step25-prepare-tsv-file-for-differential-representation-analysis">Step2.5 Prepare tsv file for differential representation analysis</a></li> </ol> </li> <li><a href="#step3perform-pairwise-comparisons-on-gene-level" id="markdown-toc-step3perform-pairwise-comparisons-on-gene-level">Step3.Perform pairwise comparisons on Gene level</a></li> <li><a href="#step4perform-pairwise-comparisons-on-shrna-level" id="markdown-toc-step4perform-pairwise-comparisons-on-shrna-level">Step4.Perform pairwise comparisons on shRNA level</a></li> </ol> <h2 id="step1mapping-to-library"> <a href="#step1mapping-to-library" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step1.Mapping to library </h2> <h3 id="step11-prepare-raw-fastq-and-library-files"> <a href="#step11-prepare-raw-fastq-and-library-files" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step1.1 Prepare raw fastq and library files </h3> <p>First of all, define your project name and create a project R object. This could be achieved by function <code class="highlighter-rouge">ScreenBEAM.dir.create</code>.</p> <p><strong>note: Project object should be library based! One library per run of ScreenBEAM2!</strong></p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">lib.name</span><span class="o">&lt;-</span><span class="s1">'[your_library_name_+_project_name]'</span><span class="w">
</span><span class="n">analysis.par</span><span class="o">&lt;-</span><span class="n">ScreenBEAM.dir.create</span><span class="p">(</span><span class="n">project_main_dir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'./'</span><span class="p">,</span><span class="w"> </span><span class="n">lib_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lib.name</span><span class="p">,</span><span class="w"> </span><span class="n">DATE</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">)</span><span class="w">

</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">par.path</span><span class="o">&lt;-</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir</span><span class="w">
</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">par.name</span><span class="o">&lt;-</span><span class="n">paste0</span><span class="p">(</span><span class="s2">"analysis_par_"</span><span class="p">,</span><span class="n">lib.name</span><span class="p">,</span><span class="w"> </span><span class="s2">".RData"</span><span class="p">)</span><span class="w">

</span><span class="n">load</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="n">par.path</span><span class="p">,</span><span class="w"> </span><span class="n">par.name</span><span class="p">))</span><span class="w">
</span></code></pre></div></div> <p>Create a table with all raw fastq files that are mapped to this library</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">f</span><span class="o">&lt;-</span><span class="n">list.files</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">"../201901/rawdata/Das_fastq/Lib_B/"</span><span class="p">,</span><span class="n">patter</span><span class="o">=</span><span class="s2">".fastq$"</span><span class="p">,</span><span class="n">recursive</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="n">full.names</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">f</span><span class="o">&lt;-</span><span class="n">normalizePath</span><span class="p">(</span><span class="n">f</span><span class="p">)</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">f</span><span class="p">,</span><span class="n">file</span><span class="o">=</span><span class="n">file.path</span><span class="p">(</span><span class="n">par.path</span><span class="p">,</span><span class="s2">"fastq_paths.txt"</span><span class="p">),</span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">,</span><span class="n">quote</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="n">row.names</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="n">col.names</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w">

</span></code></pre></div></div> <p>Next, soft link your fastq file to <code class="highlighter-rouge">analysis.par$out.dir.fastq</code> folder. This step is <strong>ESSENTIAL</strong>, please <strong>DON’T</strong> move files directly, in case of losing your data. You can go to your <code class="highlighter-rouge">analysis.par$out.dir.fastq</code> folder, then run command:</p><pre><code class="language-Shell">ln -s filepaths
</code></pre><p>Put library csv file with file name “JJ_Das_LibB.csv” in the <code class="highlighter-rouge">analysis.par$out.dir.library</code> folder. The 1st column must be gRNA name and 2nd column must be sequence. Then go to <code class="highlighter-rouge">analysis.par$out.dir.library</code> folder and run following command to create fasta file for library.</p><pre><code class="language-Shell">awk FNR-1 your_library.csv | awk -F "," '{print "&gt;"$1"\n"$2}' &gt; your_library.fa" 
</code></pre><p>Then you could move forward to mapping and collecting raw counts step, which could be executed on cluster.</p> <h3 id="step12-mapping-and-collecting-raw-counts"> <a href="#step12-mapping-and-collecting-raw-counts" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step1.2 Mapping and collecting raw counts </h3> <p>first of all, please ensure <code class="highlighter-rouge">blat</code> was installed in your working environment, since we will be utilizing <code class="highlighter-rouge">blat</code> for mapping. (You can use blat version here on conda)</p> <p>Next, we wil need to execute <code class="highlighter-rouge">ScreenBEAM.raw.count</code> function in ScreenBEAM2 R package. This function will help mapping your fastq files to library and collect raw counts by calling BLAT from R.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="w">
</span><span class="n">analysis.par</span><span class="o">&lt;-</span><span class="n">ScreenBEAM.raw.count</span><span class="p">(</span><span class="n">analysis.par</span><span class="p">)</span><span class="w">

</span><span class="c1"># suggest to save analysis.par now, because a lot of data has been created.</span><span class="w">
</span><span class="n">save</span><span class="p">(</span><span class="n">analysis.par</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="n">par.path</span><span class="p">,</span><span class="w"> </span><span class="n">par.name</span><span class="p">))</span><span class="w">

</span></code></pre></div></div> <p><strong>NOTE: Some fastq files are ultra big, you may need to chop them into small fastq files in order to execute them successfully.</strong></p> <h3 id="step13-mapping-qualtiy-control"> <a href="#step13-mapping-qualtiy-control" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step1.3 Mapping qualtiy control </h3> <p>After mapping is done, please proceed your analysis back in interactive environment (such as R studio), and run command as follows:</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ScreenBEAM.mapping.QC</span><span class="p">(</span><span class="n">analysis.par</span><span class="p">)</span><span class="w">
</span></code></pre></div></div> <p>This will output a QC report for your library mapping, including mapping rate, mismatch rate, count boxplots, etc.</p> <h2 id="step2-data-annotation-and-data-cleanning"> <a href="#step2-data-annotation-and-data-cleanning" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step2 Data annotation and data cleanning </h2> <h3 id="step21-create-metadata"> <a href="#step21-create-metadata" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step2.1 Create Metadata </h3> <p>First of all, please have your Metadata information collected in an excel file, and saved it to <code class="highlighter-rouge">analysis.par$out.dir.metadata</code>. You could take following excel file as an example. Please note that, ID that could be matched to sampleID on fastq files is required,as well as proper group informations.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">meta</span><span class="w"> </span><span class="o">&lt;-</span><span class="n">read.xlsx</span><span class="p">(</span><span class="n">file.path</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.metadata</span><span class="p">,</span><span class="s2">"Metadata.xlsx"</span><span class="p">),</span><span class="w"> </span><span class="n">sheet</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w">
</span><span class="nf">dim</span><span class="p">(</span><span class="n">meta</span><span class="p">)</span><span class="w">
</span><span class="n">raw</span><span class="o">&lt;-</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">raw.count.table</span><span class="w">
</span><span class="n">s.cur</span><span class="o">&lt;-</span><span class="n">data.frame</span><span class="p">(</span><span class="n">sampleID.full</span><span class="o">=</span><span class="nf">names</span><span class="p">(</span><span class="n">raw</span><span class="p">))</span><span class="w">
</span><span class="c1">#chop sample name into sectors</span><span class="w">
</span><span class="n">s.cur</span><span class="o">&lt;-</span><span class="n">mutate</span><span class="p">(</span><span class="n">s.cur</span><span class="p">,</span><span class="w">
              </span><span class="n">sampleID</span><span class="o">=</span><span class="n">gsub</span><span class="p">(</span><span class="s1">'(.*)_(.*)_(.*)_(.*)_R1_001'</span><span class="p">,</span><span class="s1">'\\2'</span><span class="p">,</span><span class="n">sampleID.full</span><span class="p">),</span><span class="w">
              </span><span class="n">HartwellID</span><span class="o">=</span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">gsub</span><span class="p">(</span><span class="s1">'(.*)_(.*)_(.*)_(.*)_R1_001'</span><span class="p">,</span><span class="s1">'\\1'</span><span class="p">,</span><span class="n">sampleID.full</span><span class="p">)),</span><span class="w">
              </span><span class="n">HiseqSampleID</span><span class="o">=</span><span class="n">gsub</span><span class="p">(</span><span class="s1">'(.*)_(.*)_(.*)_(.*)_R1_001'</span><span class="p">,</span><span class="s1">'\\3'</span><span class="p">,</span><span class="n">sampleID.full</span><span class="p">),</span><span class="w">
              </span><span class="n">HiseqLaneID</span><span class="o">=</span><span class="n">gsub</span><span class="p">(</span><span class="s1">'(.*)_(.*)_(.*)_(.*)_R1_001'</span><span class="p">,</span><span class="s1">'\\4'</span><span class="p">,</span><span class="n">sampleID.full</span><span class="p">)</span><span class="w">
</span><span class="p">)</span><span class="c1">#this could be adjusted accordingly</span><span class="w">

</span><span class="n">meta</span><span class="o">&lt;-</span><span class="n">left_join</span><span class="p">(</span><span class="n">s.cur</span><span class="p">,</span><span class="n">meta</span><span class="p">,</span><span class="n">by</span><span class="o">=</span><span class="s1">'sampleID'</span><span class="p">)</span><span class="w">
</span><span class="nf">dim</span><span class="p">(</span><span class="n">meta</span><span class="p">)</span><span class="w">
</span></code></pre></div></div> <p>Check if there are duplicates in your metadata – this step is crucial because duplicates will affect following data manipulation.</p> <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>meta&lt;-mutate(meta,sampleName=paste(sampleID,HiseqLaneID,sep='_'),index=paste(HiseqSampleID,HiseqLaneID,sep='_'))
filter(meta,duplicated(sampleName))
filter(meta,duplicated(index))
dim(meta)
</code></pre></div></div> <h3 id="step22-combine-mapping-info-to-your-metadata"> <a href="#step22-combine-mapping-info-to-your-metadata" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step2.2 Combine mapping info to your metadata </h3> <p>Read mapping info and combine them in meta data:</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">mapping</span><span class="o">&lt;-</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">raw.summary</span><span class="w">
</span><span class="n">mapping</span><span class="o">&lt;-</span><span class="n">mapping</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s1">'sample'</span><span class="p">,</span><span class="s1">'total'</span><span class="p">,</span><span class="s1">'n.matched'</span><span class="p">)]</span><span class="w">
</span><span class="nf">names</span><span class="p">(</span><span class="n">mapping</span><span class="p">)[</span><span class="m">1</span><span class="p">]</span><span class="o">&lt;-</span><span class="s1">'sampleID.full'</span><span class="w">
</span><span class="n">mapping</span><span class="o">&lt;-</span><span class="n">mutate</span><span class="p">(</span><span class="n">mapping</span><span class="p">,</span><span class="n">mappingRate</span><span class="o">=</span><span class="n">n.matched</span><span class="o">/</span><span class="n">total</span><span class="p">,</span><span class="n">coverage.seq</span><span class="o">=</span><span class="nf">round</span><span class="p">(</span><span class="n">n.matched</span><span class="o">/</span><span class="n">nrow</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">raw.count.table</span><span class="p">)))</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">mapping</span><span class="p">)</span><span class="w"> </span><span class="c1"># calculate mapping statistics</span><span class="w">
</span><span class="n">meta</span><span class="o">&lt;-</span><span class="n">left_join</span><span class="p">(</span><span class="n">meta</span><span class="p">,</span><span class="n">mapping</span><span class="p">,</span><span class="n">by</span><span class="o">=</span><span class="s2">"sampleID.full"</span><span class="p">)</span><span class="w">
</span><span class="n">unlink</span><span class="p">(</span><span class="n">file.path</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.mapping</span><span class="p">,</span><span class="s1">'mapping.summary.xlsx'</span><span class="p">))</span><span class="w">
</span><span class="n">write.xlsx</span><span class="p">(</span><span class="n">meta</span><span class="p">,</span><span class="n">file</span><span class="o">=</span><span class="n">file.path</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.mapping</span><span class="p">,</span><span class="s1">'mapping.summary.xlsx'</span><span class="p">),</span><span class="n">sheet</span><span class="o">=</span><span class="s1">'mapping'</span><span class="p">)</span><span class="w">
</span></code></pre></div></div> <p>Add n.mismatch filter and Save eset raw data</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">count.dist</span><span class="o">&lt;-</span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">raw.count.dist</span><span class="p">)</span><span class="w">
</span><span class="n">count.dist</span><span class="o">$</span><span class="n">sample</span><span class="o">&lt;-</span><span class="n">meta</span><span class="o">$</span><span class="n">sampleName</span><span class="p">[</span><span class="n">match</span><span class="p">(</span><span class="n">count.dist</span><span class="o">$</span><span class="n">sample</span><span class="p">,</span><span class="n">meta</span><span class="o">$</span><span class="n">sampleID.full</span><span class="p">)]</span><span class="w">
</span><span class="nf">names</span><span class="p">(</span><span class="n">count.dist</span><span class="p">)[</span><span class="m">2</span><span class="p">]</span><span class="o">&lt;-</span><span class="s1">'sampleName'</span><span class="w">
</span></code></pre></div></div> <h3 id="step23-data-normalization-and-create-expressionset"> <a href="#step23-data-normalization-and-create-expressionset" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step2.3 Data normalization and create Expressionset </h3> <p>For this step you will need to perform data normalization and store different normalization method into eset.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">normalize.total</span><span class="o">&lt;</span><span class="m">-1e6</span><span class="w">

</span><span class="n">m</span><span class="o">&lt;-</span><span class="nf">list</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="n">meta</span><span class="p">,</span><span class="n">features</span><span class="o">=</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">lib</span><span class="p">,</span><span class="w"> </span><span class="n">count.dist</span><span class="o">=</span><span class="n">count.dist</span><span class="p">)</span><span class="w">

</span><span class="n">save</span><span class="p">(</span><span class="n">m</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="n">par.path</span><span class="p">,</span><span class="w"> </span><span class="s2">"raw.list.Rdata"</span><span class="p">))</span><span class="w">
</span><span class="n">saveCountEset</span><span class="p">(</span><span class="n">m</span><span class="p">,</span><span class="w">
	</span><span class="n">save.path</span><span class="o">=</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.mapping</span><span class="p">,</span><span class="w"> 
	</span><span class="n">n.mismatch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">9</span><span class="p">)</span><span class="w"> </span><span class="c1"># n.mismatch should be adjusted according to mapping qc</span><span class="w">

</span></code></pre></div></div> <h3 id="step24-qualtiy-control-for-wrapped-expressionsets"> <a href="#step24-qualtiy-control-for-wrapped-expressionsets" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step2.4 Qualtiy control for wrapped expressionsets </h3> <p>Here we have a build-in QC function <code class="highlighter-rouge">draw.eset.QC</code> to facilitate quality control purpose. Each <code class="highlighter-rouge">draw.eset.QC</code> function will output a quality control report for following expression set.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">load</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.mapping</span><span class="p">,</span><span class="s2">"count.9mm.raw.eset"</span><span class="p">));</span><span class="w">
</span><span class="n">raw.9mm</span><span class="o">&lt;-</span><span class="n">count.Nmm.raw.eset</span><span class="w"> </span><span class="c1"># assign your eset with a more specific name</span><span class="w">
</span><span class="n">draw.eset.QC</span><span class="p">(</span><span class="n">raw.9mm</span><span class="p">,</span><span class="w"> </span><span class="n">outdir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.QC</span><span class="p">,</span><span class="w"> </span><span class="n">intgroup</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'group'</span><span class="p">,</span><span class="w"> </span><span class="n">do.logtransform</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">prefix</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'raw.9mm_'</span><span class="p">,</span><span class="w">
             </span><span class="n">choose_plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"heatmap"</span><span class="p">,</span><span class="w"> </span><span class="s2">"pca"</span><span class="p">,</span><span class="s2">"density"</span><span class="p">,</span><span class="s2">"correlation"</span><span class="p">,</span><span class="s2">"meansd"</span><span class="p">))</span><span class="w">

</span><span class="n">load</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.mapping</span><span class="p">,</span><span class="s2">"count.9mm.normalized.eset.1M.eset"</span><span class="p">));</span><span class="w">
</span><span class="n">norm.9mm</span><span class="o">&lt;-</span><span class="n">count.Nmm.normalized.eset</span><span class="w">
</span><span class="n">draw.eset.QC</span><span class="p">(</span><span class="n">norm.9mm</span><span class="p">,</span><span class="w"> </span><span class="n">outdir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.QC</span><span class="p">,</span><span class="w"> </span><span class="n">intgroup</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'group'</span><span class="p">,</span><span class="w"> </span><span class="n">do.logtransform</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">prefix</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'norm.9mm_'</span><span class="p">,</span><span class="w">
             </span><span class="n">choose_plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"heatmap"</span><span class="p">,</span><span class="w"> </span><span class="s2">"pca"</span><span class="p">,</span><span class="s2">"density"</span><span class="p">,</span><span class="s2">"correlation"</span><span class="p">,</span><span class="s2">"meansd"</span><span class="p">))</span><span class="w">

</span><span class="n">load</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.mapping</span><span class="p">,</span><span class="s2">"count.maxmm.raw.eset"</span><span class="p">));</span><span class="w">
</span><span class="n">raw.maxmm</span><span class="o">&lt;-</span><span class="n">count.maxmm.raw.eset</span><span class="w">
</span><span class="n">draw.eset.QC</span><span class="p">(</span><span class="n">raw.maxmm</span><span class="p">,</span><span class="w"> </span><span class="n">outdir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.QC</span><span class="p">,</span><span class="w"> </span><span class="n">intgroup</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'group'</span><span class="p">,</span><span class="w"> </span><span class="n">do.logtransform</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">prefix</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'raw.maxmm_'</span><span class="p">,</span><span class="w">
             </span><span class="n">choose_plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"heatmap"</span><span class="p">,</span><span class="w"> </span><span class="s2">"pca"</span><span class="p">,</span><span class="s2">"density"</span><span class="p">,</span><span class="s2">"correlation"</span><span class="p">,</span><span class="s2">"meansd"</span><span class="p">))</span><span class="w">

</span><span class="n">load</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.mapping</span><span class="p">,</span><span class="s2">"count.maxmm.normalized.1M.eset"</span><span class="p">));</span><span class="w">
</span><span class="n">norm.maxmm</span><span class="o">&lt;-</span><span class="n">count.maxmm.normalized.eset</span><span class="w">
</span><span class="n">draw.eset.QC</span><span class="p">(</span><span class="n">norm.maxmm</span><span class="p">,</span><span class="w"> </span><span class="n">outdir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.QC</span><span class="p">,</span><span class="w"> </span><span class="n">intgroup</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'group'</span><span class="p">,</span><span class="w"> </span><span class="n">do.logtransform</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">prefix</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'norm.maxmm_'</span><span class="p">,</span><span class="w">
             </span><span class="n">choose_plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"heatmap"</span><span class="p">,</span><span class="w"> </span><span class="s2">"pca"</span><span class="p">,</span><span class="s2">"density"</span><span class="p">,</span><span class="s2">"correlation"</span><span class="p">,</span><span class="s2">"meansd"</span><span class="p">))</span><span class="w">

</span></code></pre></div></div> <h3 id="step25-prepare-tsv-file-for-differential-representation-analysis"> <a href="#step25-prepare-tsv-file-for-differential-representation-analysis" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step2.5 Prepare tsv file for differential representation analysis </h3> <p>Save normalized data and ids (gene/shRNA) in a tsv file for downstream analysis.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">first.2.column</span><span class="o">&lt;-</span><span class="n">data.frame</span><span class="p">(</span><span class="n">RNAid</span><span class="o">=</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">lib</span><span class="o">$</span><span class="n">id</span><span class="p">,</span><span class="w"> </span><span class="n">geneid</span><span class="o">=</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">lib</span><span class="o">$</span><span class="n">gene</span><span class="p">)</span><span class="w">
</span><span class="n">count.column</span><span class="o">&lt;-</span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">exprs</span><span class="p">(</span><span class="n">norm.9mm</span><span class="p">))</span><span class="w">
</span><span class="n">count.column</span><span class="o">$</span><span class="n">RNAid</span><span class="o">&lt;-</span><span class="n">rownames</span><span class="p">(</span><span class="n">count.column</span><span class="p">)</span><span class="w">

</span><span class="n">final.table</span><span class="o">&lt;-</span><span class="n">merge</span><span class="p">(</span><span class="n">first.2.column</span><span class="p">,</span><span class="w"> </span><span class="n">count.column</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="s2">"RNAid"</span><span class="p">)</span><span class="w">

</span><span class="n">colnames</span><span class="p">(</span><span class="n">final.table</span><span class="p">)</span><span class="o">&lt;-</span><span class="nf">c</span><span class="p">(</span><span class="s2">"rnaID"</span><span class="p">,</span><span class="s2">"geneID"</span><span class="p">,</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="n">meta</span><span class="o">$</span><span class="n">group</span><span class="p">,</span><span class="n">meta</span><span class="o">$</span><span class="n">replicate</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">))</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">final.table</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.DR</span><span class="p">,</span><span class="s2">"YOUR_PROJECT_9mm_normalized.tsv"</span><span class="p">),</span><span class="w"> </span><span class="n">quote</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s1">'\t'</span><span class="p">)</span><span class="w">

</span></code></pre></div></div> <h2 id="step3perform-pairwise-comparisons-on-gene-level"> <a href="#step3perform-pairwise-comparisons-on-gene-level" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step3.Perform pairwise comparisons on Gene level </h2> <p>For whole genome libraries, it may cost sometime to run these pairwise comparisons sequentially. So we recommend to generate one R script for each comparisons and running them in parallel.</p> <h2 id="step4perform-pairwise-comparisons-on-shrna-level"> <a href="#step4perform-pairwise-comparisons-on-shrna-level" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step4.Perform pairwise comparisons on shRNA level </h2> <p>shRNA level comparison also give another layer of information when gene level comparisons are not that informative. Statistical significance could only be computed when there are 2 or more replicates in experinment design, otherwise only folder changes (FC) will be computed successfully.</p> <p>First of all, double check the group label in your comparisons are in <code class="highlighter-rouge">eset$group</code> info.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">unique.group</span><span class="o">&lt;-</span><span class="n">unique</span><span class="p">(</span><span class="n">eset</span><span class="o">$</span><span class="n">group</span><span class="p">)</span><span class="w"> </span><span class="c1"># group label in your eset </span><span class="w">
</span><span class="n">d</span><span class="o">&lt;-</span><span class="n">openxlsx</span><span class="o">::</span><span class="n">read.xlsx</span><span class="p">(</span><span class="s2">"20191020_CRISPR_comparisons.xlsx"</span><span class="p">)</span><span class="w"> </span><span class="c1"># comparisons that you want to perform</span><span class="w">
</span><span class="nf">all</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">d</span><span class="p">)</span><span class="o">%in%</span><span class="n">unique.group</span><span class="p">)</span><span class="w"> </span><span class="c1"># make sure all group label exists in unique.group</span><span class="w">

</span><span class="n">compare.pairs</span><span class="o">&lt;-</span><span class="n">t</span><span class="p">(</span><span class="n">d</span><span class="p">)</span><span class="w"> </span><span class="c1"># your comparisons are required to be as follows</span><span class="w">
</span></code></pre></div></div> <p>Specify your input file:</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">input.file</span><span class="o">&lt;-</span><span class="s2">"4cl_combined_withlib_normalized_50M.tsv"</span><span class="w">
</span></code></pre></div></div> <p>Do comparisons by function <code class="highlighter-rouge">ScreenBEAM.rna.level</code>.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">de.rna.list</span><span class="o">&lt;-</span><span class="nf">list</span><span class="p">()</span><span class="w"> </span><span class="c1"># create an empty list to store following analysis results</span><span class="w">
</span></code></pre></div></div> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">for</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="nf">dim</span><span class="p">(</span><span class="n">compare.pairs</span><span class="p">)[</span><span class="m">2</span><span class="p">]){</span><span class="w">
  </span><span class="n">case.groupname</span><span class="o">&lt;-</span><span class="n">compare.pairs</span><span class="p">[</span><span class="m">1</span><span class="p">,</span><span class="n">i</span><span class="p">]</span><span class="w">
  </span><span class="n">control.groupname</span><span class="o">&lt;-</span><span class="n">compare.pairs</span><span class="p">[</span><span class="m">2</span><span class="p">,</span><span class="n">i</span><span class="p">]</span><span class="w">
  
  </span><span class="n">case.sample.id</span><span class="o">&lt;-</span><span class="n">which</span><span class="p">(</span><span class="n">eset</span><span class="o">$</span><span class="n">group</span><span class="o">==</span><span class="n">compare.pairs</span><span class="p">[</span><span class="m">1</span><span class="p">,</span><span class="n">i</span><span class="p">])</span><span class="w">
  </span><span class="n">control.sample.id</span><span class="o">&lt;-</span><span class="n">which</span><span class="p">(</span><span class="n">eset</span><span class="o">$</span><span class="n">group</span><span class="o">==</span><span class="n">compare.pairs</span><span class="p">[</span><span class="m">2</span><span class="p">,</span><span class="n">i</span><span class="p">])</span><span class="w">
  
  </span><span class="n">case.postfix</span><span class="o">&lt;-</span><span class="nb">LETTERS</span><span class="p">[</span><span class="n">seq</span><span class="p">(</span><span class="n">from</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">to</span><span class="o">=</span><span class="nf">length</span><span class="p">(</span><span class="n">case.sample.id</span><span class="p">))]</span><span class="w">
  </span><span class="n">control.postfix</span><span class="o">&lt;-</span><span class="nb">LETTERS</span><span class="p">[</span><span class="n">seq</span><span class="p">(</span><span class="n">from</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">to</span><span class="o">=</span><span class="nf">length</span><span class="p">(</span><span class="n">control.sample.id</span><span class="p">))]</span><span class="w">
  
  </span><span class="n">control.samples</span><span class="o">&lt;-</span><span class="n">eset</span><span class="o">$</span><span class="n">group</span><span class="p">[</span><span class="n">control.sample.id</span><span class="p">]</span><span class="w">
  </span><span class="n">case.samples</span><span class="o">&lt;-</span><span class="n">eset</span><span class="o">$</span><span class="n">group</span><span class="p">[</span><span class="n">case.sample.id</span><span class="p">]</span><span class="w">
  
  </span><span class="n">de</span><span class="o">&lt;-</span><span class="n">ScreenBEAM.rna.level</span><span class="p">(</span><span class="n">input.file</span><span class="p">,</span><span class="w"> </span><span class="n">control.samples</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">control.samples</span><span class="p">,</span><span class="w"> </span><span class="n">case.samples</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">case.samples</span><span class="p">,</span><span class="w">
                           </span><span class="n">control.groupname</span><span class="o">=</span><span class="n">control.groupname</span><span class="p">,</span><span class="w"> </span><span class="n">case.groupname</span><span class="o">=</span><span class="n">case.groupname</span><span class="p">,</span><span class="w">
                           </span><span class="n">gene.columnId</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">do.log2</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">do.normalization</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">filterLowCount</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w">
                           </span><span class="n">filterBy</span><span class="o">=</span><span class="s1">'control'</span><span class="p">,</span><span class="n">count.cutoff</span><span class="o">=</span><span class="m">8</span><span class="p">,</span><span class="n">family</span><span class="o">=</span><span class="n">gaussian</span><span class="p">,</span><span class="n">estimation.method</span><span class="o">=</span><span class="s1">'Bayesian'</span><span class="p">)</span><span class="w">
  
  </span><span class="n">compare.name</span><span class="o">&lt;-</span><span class="n">paste0</span><span class="p">(</span><span class="n">case.groupname</span><span class="p">,</span><span class="s2">".vs."</span><span class="p">,</span><span class="n">control.groupname</span><span class="p">)</span><span class="w">
  </span><span class="nf">names</span><span class="p">(</span><span class="n">de</span><span class="p">)</span><span class="o">&lt;-</span><span class="n">paste</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">de</span><span class="p">),</span><span class="w"> </span><span class="n">compare.name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"."</span><span class="p">)</span><span class="w">
  </span><span class="nf">names</span><span class="p">(</span><span class="n">de</span><span class="p">)[</span><span class="m">1</span><span class="p">]</span><span class="o">&lt;-</span><span class="s2">"rnaID"</span><span class="w">
  </span><span class="n">de.rna.list</span><span class="p">[[</span><span class="n">compare.name</span><span class="p">]]</span><span class="o">&lt;-</span><span class="n">de</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="n">save.image</span><span class="p">(</span><span class="s2">"DE.sgRNA.RData"</span><span class="p">)</span><span class="w"> </span><span class="c1">#save your data</span><span class="w">
</span></code></pre></div></div> <p>Merge your results by “rnaID” and combine feature information from fData(eset).</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">if</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">de.rna.list</span><span class="p">)</span><span class="o">&gt;</span><span class="m">1</span><span class="p">){</span><span class="w">
  </span><span class="n">DR.RNA.DF</span><span class="o">&lt;-</span><span class="n">de.rna.list</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="w">
  </span><span class="k">for</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">2</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">de.rna.list</span><span class="p">)){</span><span class="w">
    </span><span class="n">DR.RNA.DF</span><span class="o">&lt;-</span><span class="n">merge</span><span class="p">(</span><span class="n">DR.RNA.DF</span><span class="p">,</span><span class="w"> </span><span class="n">de.rna.list</span><span class="p">[[</span><span class="n">i</span><span class="p">]],</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="s2">"rnaID"</span><span class="p">,</span><span class="n">all.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
  </span><span class="p">}</span><span class="w">
</span><span class="p">}</span><span class="k">else</span><span class="p">{</span><span class="w">
  </span><span class="n">DR.RNA.DF</span><span class="o">&lt;-</span><span class="n">de.rna.list</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="n">DR.RNA.DF</span><span class="o">&lt;-</span><span class="n">merge</span><span class="p">(</span><span class="n">DR.RNA.DF</span><span class="p">,</span><span class="n">fData</span><span class="p">(</span><span class="n">eset</span><span class="p">),</span><span class="n">by.x</span><span class="o">=</span><span class="s2">"rnaID"</span><span class="p">,</span><span class="n">by.y</span><span class="o">=</span><span class="s2">"id"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div> <p>Clean up your data frame using function <code class="highlighter-rouge">select</code>. Save your analysis results.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">DR.RNA.DF.sel</span><span class="o">&lt;-</span><span class="n">dplyr</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">DR.RNA.DF</span><span class="p">,</span><span class="w">
                      </span><span class="n">rnaID</span><span class="p">,</span><span class="w">
                      </span><span class="n">gRNA.sequence</span><span class="o">:</span><span class="n">nchar</span><span class="p">,</span><span class="w">
                      </span><span class="n">dplyr</span><span class="o">::</span><span class="n">starts_with</span><span class="p">(</span><span class="s1">'log2FC'</span><span class="p">),</span><span class="w">
                      </span><span class="n">dplyr</span><span class="o">::</span><span class="n">starts_with</span><span class="p">(</span><span class="s1">'Ave.'</span><span class="p">)</span><span class="w">
</span><span class="p">)</span><span class="w">
</span><span class="n">save.image</span><span class="p">(</span><span class="s2">"DE.sgRNA.RData"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div> <p>Output your result as an excel file.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">df</span><span class="o">&lt;-</span><span class="n">exprs</span><span class="p">(</span><span class="n">eset</span><span class="p">)</span><span class="w">
</span><span class="n">DR.RNA.DF.sel</span><span class="o">&lt;-</span><span class="n">merge</span><span class="p">(</span><span class="n">DR.RNA.DF.sel</span><span class="p">,</span><span class="n">df</span><span class="p">,</span><span class="n">by.x</span><span class="o">=</span><span class="s2">"rnaID"</span><span class="p">,</span><span class="n">by.y</span><span class="o">=</span><span class="s2">"row.names"</span><span class="p">)</span><span class="w">

</span><span class="n">openxlsx</span><span class="o">::</span><span class="n">write.xlsx</span><span class="p">(</span><span class="n">DR.RNA.DF.sel</span><span class="p">,</span><span class="w"> </span><span class="s2">"4CL.2Lib.DR.sgRNA.xlsx"</span><span class="p">)</span><span class="w">

</span></code></pre></div></div> <hr> <footer role="contentinfo"> <p class="text-small text-grey-dk-000 mb-0">Copyright &copy; 2019- Yu Lab. Distributed by an <a href="https://github.com/jyyulab/scMINER/blob/master/LICENSE.txt">Apache License 2.0.</a></p> </footer> </div> </div> </div> </div> </body> </html>
