<!DOCTYPE html> <html lang="en-US"> <head> <meta charset="UTF-8"> <meta http-equiv="X-UA-Compatible" content="IE=Edge"> <title>Tutorial - Cleanning - ScreenBEAM</title> <link rel="shortcut icon" href="http://localhost:4000/favicon.ico" type="image/x-icon"> <link rel="stylesheet" href="http://localhost:4000/assets/css/just-the-docs.css"> <script type="text/javascript" src="http://localhost:4000/assets/js/vendor/lunr.min.js"></script> <script type="text/javascript" src="http://localhost:4000/assets/js/just-the-docs.js"></script> <meta name="viewport" content="width=device-width, initial-scale=1"> <!-- Begin Jekyll SEO tag v2.6.1 --> <title>Tutorial - Cleanning | ScreenBEAM</title> <meta name="generator" content="Jekyll v3.8.6" /> <meta property="og:title" content="Tutorial - Cleanning" /> <meta property="og:locale" content="en_US" /> <link rel="canonical" href="http://localhost:4000/docs/Tutorial2.html" /> <meta property="og:url" content="http://localhost:4000/docs/Tutorial2.html" /> <meta property="og:site_name" content="ScreenBEAM" /> <script type="application/ld+json"> {"url":"http://localhost:4000/docs/Tutorial2.html","headline":"Tutorial - Cleanning","@type":"WebPage","@context":"https://schema.org"}</script> <!-- End Jekyll SEO tag --> </head> <body> <svg xmlns="http://www.w3.org/2000/svg" style="display: none;"> <symbol id="link" viewBox="0 0 16 16"> <title>Link</title> <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path> </symbol> </svg> <div class="page-wrap"> <div class="side-bar"> <div class="site-header"> <a href="http://localhost:4000" class="site-title lh-tight"> ScreenBEAM </a> <button class="menu-button fs-3 js-main-nav-trigger" data-text-toggle="Hide" type="button">Menu</button> </div> <div class="navigation main-nav js-main-nav"> <nav role="navigation" aria-label="Main navigation"> <ul class="navigation-list"><li class="navigation-list-item"><a href="http://localhost:4000/" class="navigation-list-link">Home</a></li><li class="navigation-list-item"><a href="http://localhost:4000/docs/Installation.html" class="navigation-list-link">Installation</a></li><li class="navigation-list-item"><a href="http://localhost:4000/docs/Tutorial1.html" class="navigation-list-link">Tutorial - Mapping</a></li><li class="navigation-list-item active"><a href="http://localhost:4000/docs/Tutorial2.html" class="navigation-list-link active">Tutorial - Cleanning</a></li><li class="navigation-list-item"><a href="http://localhost:4000/docs/Tutorial3.html" class="navigation-list-link">Tutorial - DR</a></li><li class="navigation-list-item"><a href="http://localhost:4000/docs/hpc_tutorial.html" class="navigation-list-link">Running on cluster</a></li><li class="navigation-list-item"><a href="http://localhost:4000/docs/Acknowledgements.html" class="navigation-list-link">Acknowledgements</a></li></ul> </nav> </div> <footer class="site-footer"> <p class="text-small text-grey-dk-000 mb-4">This site uses <a href="https://github.com/pmarsceill/just-the-docs">Just the Docs</a>, a documentation theme for Jekyll.</p> </footer> </div> <div class="main-content-wrap js-main-content" tabindex="0"> <div class="main-content"> <div class="page-header js-page-header"> <div class="search"> <div class="search-input-wrap"> <input type="text" class="js-search-input search-input" tabindex="0" placeholder="Search ScreenBEAM" aria-label="Search ScreenBEAM" autocomplete="off"> <svg width="14" height="14" viewBox="0 0 28 28" xmlns="http://www.w3.org/2000/svg" class="search-icon"><title>Search</title><g fill-rule="nonzero"><path d="M17.332 20.735c-5.537 0-10-4.6-10-10.247 0-5.646 4.463-10.247 10-10.247 5.536 0 10 4.601 10 10.247s-4.464 10.247-10 10.247zm0-4c3.3 0 6-2.783 6-6.247 0-3.463-2.7-6.247-6-6.247s-6 2.784-6 6.247c0 3.464 2.7 6.247 6 6.247z"/><path d="M11.672 13.791L.192 25.271 3.02 28.1 14.5 16.62z"/></g></svg> </div> <div class="js-search-results search-results-wrap"></div> </div> </div> <div class="page"> <div id="main-content" class="page-content" role="main"> <h1 class="no_toc" id="tutorial-2-data-annotation-and-data-cleanning"> <a href="#tutorial-2-data-annotation-and-data-cleanning" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Tutorial-2 Data annotation and data cleanning </h1> <p>ScreenBEAM2 is a R based tool which consists of three major parts for processing steps: 1. mapping long read sequence to short read libraries, 2. Quality control, data cleanning and data preprocessing for mapped raw counts data; 3. Differential representative analysis on gene level or shRNA level.</p> <p><strong>This is the second part of the whole tutorial, which is focused on expression set cleaning and meta data integration.</strong></p> <h2 class="no_toc text-delta" id="table-of-contents"> <a href="#table-of-contents" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Table of contents </h2> <ol id="markdown-toc"> <li><a href="#step2-data-annotation-and-data-cleanning" id="markdown-toc-step2-data-annotation-and-data-cleanning">Step2 Data annotation and data cleanning</a> <ol> <li><a href="#step21-create-metadata" id="markdown-toc-step21-create-metadata">Step2.1 Create Metadata</a></li> <li><a href="#step22-combine-mapping-info-to-your-metadata" id="markdown-toc-step22-combine-mapping-info-to-your-metadata">Step2.2 Combine mapping info to your metadata</a></li> <li><a href="#step23-data-normalization-and-create-expressionset" id="markdown-toc-step23-data-normalization-and-create-expressionset">Step2.3 Data normalization and create Expressionset</a></li> <li><a href="#step24-qualtiy-control-for-wrapped-expressionsets" id="markdown-toc-step24-qualtiy-control-for-wrapped-expressionsets">Step2.4 Qualtiy control for wrapped expressionsets</a></li> <li><a href="#step25-prepare-tsv-file-for-differential-representation-analysis" id="markdown-toc-step25-prepare-tsv-file-for-differential-representation-analysis">Step2.5 Prepare tsv file for differential representation analysis</a></li> </ol> </li> </ol> <h2 id="step2-data-annotation-and-data-cleanning"> <a href="#step2-data-annotation-and-data-cleanning" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step2 Data annotation and data cleanning </h2> <h3 id="step21-create-metadata"> <a href="#step21-create-metadata" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step2.1 Create Metadata </h3> <p>First of all, please have your Metadata information collected in an excel file, and saved it to <code class="highlighter-rouge">analysis.par$out.dir.metadata</code>. Please note that, ID that could be matched to sampleID on fastq files is required, as well as proper group informations. A sample metadata file is shown as follows:</p> <div class="table-wrapper"><table> <thead> <tr> <th><strong>sampleID</strong></th> <th style="text-align: center"><strong>group</strong></th> <th style="text-align: right">description</th> <th style="text-align: right">replicate</th> <th style="text-align: right">expName</th> <th> </th> </tr> </thead> <tbody> <tr> <td> </td> <td style="text-align: center">F171</td> <td style="text-align: right">0.25Hi</td> <td style="text-align: right">0.25ug nacl-high</td> <td style="text-align: right">A</td> <td>2nd round screening in cells using library A</td> </tr> <tr> <td> </td> <td style="text-align: center">F172</td> <td style="text-align: right">0.25Lo</td> <td style="text-align: right">0.25ug nacl-low</td> <td style="text-align: right">A</td> <td>2nd round screening in cells using library A</td> </tr> <tr> <td> </td> <td style="text-align: center">F173</td> <td style="text-align: right">0.25Hi</td> <td style="text-align: right">0.25ug nacl-high</td> <td style="text-align: right">B</td> <td>2nd round screening in cells using library A</td> </tr> </tbody> </table></div> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">meta</span><span class="w"> </span><span class="o">&lt;-</span><span class="n">read.xlsx</span><span class="p">(</span><span class="n">file.path</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.metadata</span><span class="p">,</span><span class="s2">"Metadata.xlsx"</span><span class="p">),</span><span class="w"> </span><span class="n">sheet</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w">
</span><span class="nf">dim</span><span class="p">(</span><span class="n">meta</span><span class="p">)</span><span class="w">
</span><span class="n">raw</span><span class="o">&lt;-</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">raw.count.table</span><span class="w">
</span><span class="n">s.cur</span><span class="o">&lt;-</span><span class="n">data.frame</span><span class="p">(</span><span class="n">sampleID.full</span><span class="o">=</span><span class="nf">names</span><span class="p">(</span><span class="n">raw</span><span class="p">))</span><span class="w">

</span><span class="c1">#chop sample name into sectors</span><span class="w">
</span><span class="n">s.cur</span><span class="o">&lt;-</span><span class="n">mutate</span><span class="p">(</span><span class="n">s.cur</span><span class="p">,</span><span class="w">
              </span><span class="n">sampleID</span><span class="o">=</span><span class="n">gsub</span><span class="p">(</span><span class="s1">'(.*)_(.*)_(.*)_(.*)_R1_001'</span><span class="p">,</span><span class="s1">'\\2'</span><span class="p">,</span><span class="n">sampleID.full</span><span class="p">),</span><span class="w">
              </span><span class="n">HartwellID</span><span class="o">=</span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">gsub</span><span class="p">(</span><span class="s1">'(.*)_(.*)_(.*)_(.*)_R1_001'</span><span class="p">,</span><span class="s1">'\\1'</span><span class="p">,</span><span class="n">sampleID.full</span><span class="p">)),</span><span class="w">
              </span><span class="n">HiseqSampleID</span><span class="o">=</span><span class="n">gsub</span><span class="p">(</span><span class="s1">'(.*)_(.*)_(.*)_(.*)_R1_001'</span><span class="p">,</span><span class="s1">'\\3'</span><span class="p">,</span><span class="n">sampleID.full</span><span class="p">),</span><span class="w">
              </span><span class="n">HiseqLaneID</span><span class="o">=</span><span class="n">gsub</span><span class="p">(</span><span class="s1">'(.*)_(.*)_(.*)_(.*)_R1_001'</span><span class="p">,</span><span class="s1">'\\4'</span><span class="p">,</span><span class="n">sampleID.full</span><span class="p">)</span><span class="w">
</span><span class="p">)</span><span class="c1">#this could be adjusted according to names of your fastq files</span><span class="w">

</span><span class="n">meta</span><span class="o">&lt;-</span><span class="n">left_join</span><span class="p">(</span><span class="n">s.cur</span><span class="p">,</span><span class="n">meta</span><span class="p">,</span><span class="n">by</span><span class="o">=</span><span class="s1">'sampleID'</span><span class="p">)</span><span class="w"> </span><span class="c1"># merge sample info from fastq with your meta data</span><span class="w">
</span><span class="nf">dim</span><span class="p">(</span><span class="n">meta</span><span class="p">)</span><span class="w">
</span></code></pre></div></div> <p>Create unique sample name and index for each sample by pasting sampleID and lane ID together. This could be adjusted accordingly.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">meta</span><span class="o">&lt;-</span><span class="n">mutate</span><span class="p">(</span><span class="n">meta</span><span class="p">,</span><span class="n">sampleName</span><span class="o">=</span><span class="n">paste</span><span class="p">(</span><span class="n">sampleID</span><span class="p">,</span><span class="n">HiseqLaneID</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">'_'</span><span class="p">),</span><span class="w">  
	</span><span class="n">index</span><span class="o">=</span><span class="n">paste</span><span class="p">(</span><span class="n">HiseqSampleID</span><span class="p">,</span><span class="n">HiseqLaneID</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">'_'</span><span class="p">))</span><span class="w"> 
</span></code></pre></div></div> <p>Check if there are duplicates in your metadata – this step is crucial because duplicates will affect following data manipulation.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">filter</span><span class="p">(</span><span class="n">meta</span><span class="p">,</span><span class="n">duplicated</span><span class="p">(</span><span class="n">sampleName</span><span class="p">))</span><span class="w">
</span><span class="n">filter</span><span class="p">(</span><span class="n">meta</span><span class="p">,</span><span class="n">duplicated</span><span class="p">(</span><span class="n">index</span><span class="p">))</span><span class="w">
</span><span class="nf">dim</span><span class="p">(</span><span class="n">meta</span><span class="p">)</span><span class="w">
</span></code></pre></div></div> <h3 id="step22-combine-mapping-info-to-your-metadata"> <a href="#step22-combine-mapping-info-to-your-metadata" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step2.2 Combine mapping info to your metadata </h3> <p>After all mapping jobs are completed and mapping statistics are generated and stored in designated location <code class="highlighter-rouge">analysis.par$raw.summary</code>. We could calculate mapping rate, and sequencing coverages from mapping statistics and store them in meta data.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">mapping</span><span class="o">&lt;-</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">raw.summary</span><span class="w">
</span><span class="n">mapping</span><span class="o">&lt;-</span><span class="n">mapping</span><span class="p">[,</span><span class="nf">c</span><span class="p">(</span><span class="s1">'sample'</span><span class="p">,</span><span class="s1">'total'</span><span class="p">,</span><span class="s1">'n.matched'</span><span class="p">)]</span><span class="w">
</span><span class="nf">names</span><span class="p">(</span><span class="n">mapping</span><span class="p">)[</span><span class="m">1</span><span class="p">]</span><span class="o">&lt;-</span><span class="s1">'sampleID.full'</span><span class="w">

</span><span class="c1">#calculate mapping rate and sequencing coverages</span><span class="w">
</span><span class="n">mapping</span><span class="o">&lt;-</span><span class="n">mutate</span><span class="p">(</span><span class="n">mapping</span><span class="p">,</span><span class="n">mappingRate</span><span class="o">=</span><span class="n">n.matched</span><span class="o">/</span><span class="n">total</span><span class="p">,</span><span class="n">coverage.seq</span><span class="o">=</span><span class="nf">round</span><span class="p">(</span><span class="n">n.matched</span><span class="o">/</span><span class="n">nrow</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">raw.count.table</span><span class="p">)))</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">mapping</span><span class="p">)</span><span class="w"> 

</span><span class="c1">#combine mapping statistics with meta data, save files</span><span class="w">
</span><span class="n">meta</span><span class="o">&lt;-</span><span class="n">left_join</span><span class="p">(</span><span class="n">meta</span><span class="p">,</span><span class="n">mapping</span><span class="p">,</span><span class="n">by</span><span class="o">=</span><span class="s2">"sampleID.full"</span><span class="p">)</span><span class="w">
</span><span class="n">unlink</span><span class="p">(</span><span class="n">file.path</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.mapping</span><span class="p">,</span><span class="s1">'mapping.summary.xlsx'</span><span class="p">))</span><span class="w">
</span><span class="n">write.xlsx</span><span class="p">(</span><span class="n">meta</span><span class="p">,</span><span class="n">file</span><span class="o">=</span><span class="n">file.path</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.mapping</span><span class="p">,</span><span class="s1">'mapping.summary.xlsx'</span><span class="p">),</span><span class="n">sheet</span><span class="o">=</span><span class="s1">'mapping'</span><span class="p">)</span><span class="w">
</span></code></pre></div></div> <p>Add sample information to count distribution matrix.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">count.dist</span><span class="o">&lt;-</span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">raw.count.dist</span><span class="p">)</span><span class="w">
</span><span class="n">count.dist</span><span class="o">$</span><span class="n">sample</span><span class="o">&lt;-</span><span class="n">meta</span><span class="o">$</span><span class="n">sampleName</span><span class="p">[</span><span class="n">match</span><span class="p">(</span><span class="n">count.dist</span><span class="o">$</span><span class="n">sample</span><span class="p">,</span><span class="n">meta</span><span class="o">$</span><span class="n">sampleID.full</span><span class="p">)]</span><span class="w">
</span><span class="nf">names</span><span class="p">(</span><span class="n">count.dist</span><span class="p">)[</span><span class="m">2</span><span class="p">]</span><span class="o">&lt;-</span><span class="s1">'sampleName'</span><span class="w">
</span></code></pre></div></div> <p>Now we are ready to deal with the count data and create expression sets.</p> <h3 id="step23-data-normalization-and-create-expressionset"> <a href="#step23-data-normalization-and-create-expressionset" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step2.3 Data normalization and create Expressionset </h3> <p>Then, we could perform data normalization and store different normalization method into eset. Total number of normalization is data dependent, usually we use 1e6.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">normalize.total</span><span class="o">&lt;</span><span class="m">-1e6</span><span class="w">
</span><span class="n">m</span><span class="o">&lt;-</span><span class="nf">list</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="n">meta</span><span class="p">,</span><span class="n">features</span><span class="o">=</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">lib</span><span class="p">,</span><span class="w"> </span><span class="n">count.dist</span><span class="o">=</span><span class="n">count.dist</span><span class="p">)</span><span class="w">
</span><span class="n">save</span><span class="p">(</span><span class="n">m</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="n">par.path</span><span class="p">,</span><span class="w"> </span><span class="s2">"raw.list.Rdata"</span><span class="p">))</span><span class="w">
</span></code></pre></div></div> <p>Save your expression matrix based on number of mismatch, this optimal mismatch number could be found in previous mapping QC report, as an example here we use 9. This function will also save an expression set with maximum mismatch and 0 mismatch, for your reference.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">saveCountEset</span><span class="p">(</span><span class="n">m</span><span class="p">,</span><span class="w">
	</span><span class="n">save.path</span><span class="o">=</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.mapping</span><span class="p">,</span><span class="w"> 
	</span><span class="n">n.mismatch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">9</span><span class="p">)</span><span class="w"> </span><span class="c1"># n.mismatch should be adjusted according to mapping qc</span><span class="w">
</span></code></pre></div></div> <h3 id="step24-qualtiy-control-for-wrapped-expressionsets"> <a href="#step24-qualtiy-control-for-wrapped-expressionsets" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step2.4 Qualtiy control for wrapped expressionsets </h3> <p>Here we have a build-in QC function imported from NetBID2 – <code class="highlighter-rouge">draw.eset.QC</code> to facilitate quality control purpose. Each <code class="highlighter-rouge">draw.eset.QC</code> function will output a quality control report for following expression set.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">load</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.mapping</span><span class="p">,</span><span class="s2">"count.9mm.raw.eset"</span><span class="p">));</span><span class="w">
</span><span class="n">raw.9mm</span><span class="o">&lt;-</span><span class="n">count.Nmm.raw.eset</span><span class="w"> </span><span class="c1"># assign your eset with a more specific name</span><span class="w">
</span><span class="n">draw.eset.QC</span><span class="p">(</span><span class="n">raw.9mm</span><span class="p">,</span><span class="w"> </span><span class="n">outdir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.QC</span><span class="p">,</span><span class="w"> </span><span class="n">intgroup</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'group'</span><span class="p">,</span><span class="w"> </span><span class="n">do.logtransform</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">prefix</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'raw.9mm_'</span><span class="p">,</span><span class="w"> </span><span class="n">choose_plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"heatmap"</span><span class="p">,</span><span class="w"> </span><span class="s2">"pca"</span><span class="p">,</span><span class="s2">"density"</span><span class="p">,</span><span class="s2">"correlation"</span><span class="p">,</span><span class="s2">"meansd"</span><span class="p">))</span><span class="w">
</span></code></pre></div></div> <p>Here we could also choose to perform more quality control for different numbers of mismatch, as a reference.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">load</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.mapping</span><span class="p">,</span><span class="s2">"count.9mm.normalized.eset.1M.eset"</span><span class="p">));</span><span class="w">
</span><span class="n">norm.9mm</span><span class="o">&lt;-</span><span class="n">count.Nmm.normalized.eset</span><span class="w">
</span><span class="n">draw.eset.QC</span><span class="p">(</span><span class="n">norm.9mm</span><span class="p">,</span><span class="w"> </span><span class="n">outdir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.QC</span><span class="p">,</span><span class="w"> </span><span class="n">intgroup</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'group'</span><span class="p">,</span><span class="w"> </span><span class="n">do.logtransform</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">prefix</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'norm.9mm_'</span><span class="p">,</span><span class="w"> </span><span class="n">choose_plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"heatmap"</span><span class="p">,</span><span class="w"> </span><span class="s2">"pca"</span><span class="p">,</span><span class="s2">"density"</span><span class="p">,</span><span class="s2">"correlation"</span><span class="p">,</span><span class="s2">"meansd"</span><span class="p">))</span><span class="w">

</span><span class="n">load</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.mapping</span><span class="p">,</span><span class="s2">"count.maxmm.raw.eset"</span><span class="p">));</span><span class="w">
</span><span class="n">raw.maxmm</span><span class="o">&lt;-</span><span class="n">count.maxmm.raw.eset</span><span class="w">
</span><span class="n">draw.eset.QC</span><span class="p">(</span><span class="n">raw.maxmm</span><span class="p">,</span><span class="w"> </span><span class="n">outdir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.QC</span><span class="p">,</span><span class="w"> </span><span class="n">intgroup</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'group'</span><span class="p">,</span><span class="w"> </span><span class="n">do.logtransform</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">prefix</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'raw.maxmm_'</span><span class="p">,</span><span class="w"> </span><span class="n">choose_plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"heatmap"</span><span class="p">,</span><span class="w"> </span><span class="s2">"pca"</span><span class="p">,</span><span class="s2">"density"</span><span class="p">,</span><span class="s2">"correlation"</span><span class="p">,</span><span class="s2">"meansd"</span><span class="p">))</span><span class="w">

</span><span class="n">load</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.mapping</span><span class="p">,</span><span class="s2">"count.maxmm.normalized.1M.eset"</span><span class="p">));</span><span class="w">
</span><span class="n">norm.maxmm</span><span class="o">&lt;-</span><span class="n">count.maxmm.normalized.eset</span><span class="w">
</span><span class="n">draw.eset.QC</span><span class="p">(</span><span class="n">norm.maxmm</span><span class="p">,</span><span class="w"> </span><span class="n">outdir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.QC</span><span class="p">,</span><span class="w"> </span><span class="n">intgroup</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'group'</span><span class="p">,</span><span class="w"> </span><span class="n">do.logtransform</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">prefix</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'norm.maxmm_'</span><span class="p">,</span><span class="w"> </span><span class="n">choose_plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"heatmap"</span><span class="p">,</span><span class="w"> </span><span class="s2">"pca"</span><span class="p">,</span><span class="s2">"density"</span><span class="p">,</span><span class="s2">"correlation"</span><span class="p">,</span><span class="s2">"meansd"</span><span class="p">))</span><span class="w">
</span></code></pre></div></div> <h3 id="step25-prepare-tsv-file-for-differential-representation-analysis"> <a href="#step25-prepare-tsv-file-for-differential-representation-analysis" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step2.5 Prepare tsv file for differential representation analysis </h3> <p>Save normalized data and ids (gene/shRNA) in a tsv file for downstream analysis.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">first.2.column</span><span class="o">&lt;-</span><span class="n">data.frame</span><span class="p">(</span><span class="n">RNAid</span><span class="o">=</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">lib</span><span class="o">$</span><span class="n">id</span><span class="p">,</span><span class="w"> </span><span class="n">geneid</span><span class="o">=</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">lib</span><span class="o">$</span><span class="n">gene</span><span class="p">)</span><span class="w">
</span><span class="n">count.column</span><span class="o">&lt;-</span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">exprs</span><span class="p">(</span><span class="n">norm.9mm</span><span class="p">))</span><span class="w">
</span><span class="n">count.column</span><span class="o">$</span><span class="n">RNAid</span><span class="o">&lt;-</span><span class="n">rownames</span><span class="p">(</span><span class="n">count.column</span><span class="p">)</span><span class="w">

</span><span class="n">final.table</span><span class="o">&lt;-</span><span class="n">merge</span><span class="p">(</span><span class="n">first.2.column</span><span class="p">,</span><span class="w"> </span><span class="n">count.column</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="s2">"RNAid"</span><span class="p">)</span><span class="w">

</span><span class="n">colnames</span><span class="p">(</span><span class="n">final.table</span><span class="p">)</span><span class="o">&lt;-</span><span class="nf">c</span><span class="p">(</span><span class="s2">"rnaID"</span><span class="p">,</span><span class="s2">"geneID"</span><span class="p">,</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="n">meta</span><span class="o">$</span><span class="n">group</span><span class="p">,</span><span class="n">meta</span><span class="o">$</span><span class="n">replicate</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">))</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">final.table</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir.output.DR</span><span class="p">,</span><span class="s2">"YOUR_PROJECT_9mm_normalized.tsv"</span><span class="p">),</span><span class="w"> </span><span class="n">quote</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s1">'\t'</span><span class="p">)</span><span class="w">

</span></code></pre></div></div> <p>After writing table, you will be able to quit current working environment and run following analysis on cluster. Or if your data size is moderate, you could stay in your r environment and move forward.</p> <hr> <footer role="contentinfo"> <p class="text-small text-grey-dk-000 mb-0">Copyright &copy; 2019- Yu Lab. Distributed by an <a href="https://github.com/jyyulab/scMINER/blob/master/LICENSE.txt">Apache License 2.0.</a></p> </footer> </div> </div> </div> </div> </body> </html>
