<!DOCTYPE html> <html lang="en-US"> <head> <meta charset="UTF-8"> <meta http-equiv="X-UA-Compatible" content="IE=Edge"> <title>1.Mapping - ScreenBEAM</title> <link rel="shortcut icon" href="http://localhost:4000/favicon.ico" type="image/x-icon"> <link rel="stylesheet" href="http://localhost:4000/assets/css/just-the-docs.css"> <script type="text/javascript" src="http://localhost:4000/assets/js/vendor/lunr.min.js"></script> <script type="text/javascript" src="http://localhost:4000/assets/js/just-the-docs.js"></script> <meta name="viewport" content="width=device-width, initial-scale=1"> <!-- Begin Jekyll SEO tag v2.6.1 --> <title>1.Mapping | ScreenBEAM</title> <meta name="generator" content="Jekyll v3.8.6" /> <meta property="og:title" content="1.Mapping" /> <meta property="og:locale" content="en_US" /> <link rel="canonical" href="http://localhost:4000/docs/Tutorial1.html" /> <meta property="og:url" content="http://localhost:4000/docs/Tutorial1.html" /> <meta property="og:site_name" content="ScreenBEAM" /> <script type="application/ld+json"> {"url":"http://localhost:4000/docs/Tutorial1.html","headline":"1.Mapping","@type":"WebPage","@context":"https://schema.org"}</script> <!-- End Jekyll SEO tag --> </head> <body> <svg xmlns="http://www.w3.org/2000/svg" style="display: none;"> <symbol id="link" viewBox="0 0 16 16"> <title>Link</title> <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path> </symbol> </svg> <div class="page-wrap"> <div class="side-bar"> <div class="site-header"> <a href="http://localhost:4000" class="site-title lh-tight"> ScreenBEAM </a> <button class="menu-button fs-3 js-main-nav-trigger" data-text-toggle="Hide" type="button">Menu</button> </div> <div class="navigation main-nav js-main-nav"> <nav role="navigation" aria-label="Main navigation"> <ul class="navigation-list"><li class="navigation-list-item"><a href="http://localhost:4000/" class="navigation-list-link">Home</a></li><li class="navigation-list-item"><a href="http://localhost:4000/docs/Installation.html" class="navigation-list-link">Installation</a></li><li class="navigation-list-item active"><a href="http://localhost:4000/docs/Guided_analysis.html" class="navigation-list-link">Guided tutorials</a><ul class="navigation-list-child-list "><li class="navigation-list-item active"><a href="http://localhost:4000/docs/Tutorial1.html" class="navigation-list-link active">1.Mapping</a></li><li class="navigation-list-item "><a href="http://localhost:4000/docs/Tutorial2.html" class="navigation-list-link">2.Cleanning</a></li><li class="navigation-list-item "><a href="http://localhost:4000/docs/Tutorial3.html" class="navigation-list-link">3.DR</a></li></ul></li><li class="navigation-list-item"><a href="http://localhost:4000/docs/hpc_tutorial.html" class="navigation-list-link">Running on cluster</a></li><li class="navigation-list-item"><a href="http://localhost:4000/docs/Acknowledgements.html" class="navigation-list-link">Acknowledgements</a></li></ul> </nav> </div> <footer class="site-footer"> <p class="text-small text-grey-dk-000 mb-4">This site uses <a href="https://github.com/pmarsceill/just-the-docs">Just the Docs</a>, a documentation theme for Jekyll.</p> </footer> </div> <div class="main-content-wrap js-main-content" tabindex="0"> <div class="main-content"> <div class="page-header js-page-header"> <div class="search"> <div class="search-input-wrap"> <input type="text" class="js-search-input search-input" tabindex="0" placeholder="Search ScreenBEAM" aria-label="Search ScreenBEAM" autocomplete="off"> <svg width="14" height="14" viewBox="0 0 28 28" xmlns="http://www.w3.org/2000/svg" class="search-icon"><title>Search</title><g fill-rule="nonzero"><path d="M17.332 20.735c-5.537 0-10-4.6-10-10.247 0-5.646 4.463-10.247 10-10.247 5.536 0 10 4.601 10 10.247s-4.464 10.247-10 10.247zm0-4c3.3 0 6-2.783 6-6.247 0-3.463-2.7-6.247-6-6.247s-6 2.784-6 6.247c0 3.464 2.7 6.247 6 6.247z"/><path d="M11.672 13.791L.192 25.271 3.02 28.1 14.5 16.62z"/></g></svg> </div> <div class="js-search-results search-results-wrap"></div> </div> </div> <div class="page"> <nav class="breadcrumb-nav"> <ol class="breadcrumb-nav-list"> <li class="breadcrumb-nav-list-item"><a href="http://localhost:4000/docs/Guided_analysis.html">Guided tutorials</a></li> <li class="breadcrumb-nav-list-item"><span>1.Mapping</span></li> </ol> </nav> <div id="main-content" class="page-content" role="main"> <h1 class="no_toc" id="tutorial---mapping-via-screenbeam2"> <a href="#tutorial---mapping-via-screenbeam2" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Tutorial - Mapping via ScreenBEAM2 </h1> <p>ScreenBEAM2 is a R based tool which consists of three major parts for processing steps: 1. mapping long read sequence to short read libraries, 2. Quality control, data cleanning and data preprocessing for mapped raw counts data; 3. Differential representative analysis on gene level or shRNA level.</p> <p><strong>This is the first part of the whole tutorial, which is focused on mapping your long read fastq file with short read libraries.</strong></p> <h2 class="no_toc text-delta" id="table-of-contents"> <a href="#table-of-contents" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Table of contents </h2> <ol id="markdown-toc"> <li><a href="#step1mapping-to-library" id="markdown-toc-step1mapping-to-library">Step1.Mapping to library</a> <ol> <li><a href="#step11-prepare-raw-fastq-and-library-files" id="markdown-toc-step11-prepare-raw-fastq-and-library-files">Step1.1 Prepare raw fastq and library files</a></li> <li><a href="#step12-mapping-and-collecting-raw-counts" id="markdown-toc-step12-mapping-and-collecting-raw-counts">Step1.2 Mapping and collecting raw counts</a></li> <li><a href="#step13-mapping-qualtiy-control" id="markdown-toc-step13-mapping-qualtiy-control">Step1.3 Mapping qualtiy control</a></li> <li><a href="#step14-read-your-mapping-quality-control" id="markdown-toc-step14-read-your-mapping-quality-control">Step1.4 Read your mapping quality control</a></li> </ol> </li> </ol> <h2 id="step1mapping-to-library"> <a href="#step1mapping-to-library" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step1.Mapping to library </h2> <h3 id="step11-prepare-raw-fastq-and-library-files"> <a href="#step11-prepare-raw-fastq-and-library-files" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step1.1 Prepare raw fastq and library files </h3> <p>First of all, define your project name and create a project R object. This could be achieved by function <code class="highlighter-rouge">ScreenBEAM.dir.create</code>.</p> <p><strong>note: Project object should be library based! One library per run of ScreenBEAM2!</strong> If you have multiple libraries mapped at the same time, you have to open different R project for different library runs. Our only allows one library at a time.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">lib.name</span><span class="o">&lt;-</span><span class="s1">'[your_library_name_+_project_name]'</span><span class="w">
</span><span class="n">analysis.par</span><span class="o">&lt;-</span><span class="n">ScreenBEAM.dir.create</span><span class="p">(</span><span class="n">project_main_dir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'./'</span><span class="p">,</span><span class="w"> </span><span class="n">lib_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lib.name</span><span class="p">,</span><span class="w"> </span><span class="n">DATE</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">)</span><span class="w">

</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">par.path</span><span class="o">&lt;-</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir</span><span class="w"> 
</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">par.name</span><span class="o">&lt;-</span><span class="n">paste0</span><span class="p">(</span><span class="s2">"analysis_par_"</span><span class="p">,</span><span class="n">lib.name</span><span class="p">,</span><span class="w"> </span><span class="s2">".RData"</span><span class="p">)</span><span class="w">

</span><span class="n">save.image</span><span class="p">(</span><span class="n">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">par.path</span><span class="p">,</span><span class="w"> </span><span class="n">analysis.par</span><span class="o">$</span><span class="n">par.name</span><span class="p">))</span><span class="w"> </span><span class="c1">#save your image</span><span class="w">
</span></code></pre></div></div> <p>Next, soft link your fastq file to <code class="highlighter-rouge">analysis.par$out.dir.fastq</code> folder. This step is <strong>ESSENTIAL</strong>, please <strong>DON’T</strong> move files directly, in case of losing your data. Before you soft link data files, we suggested double check your fastq files with your meta data, to make sure all fastq files were mapped to your designated library.</p> <p>You can go to your <code class="highlighter-rouge">analysis.par$out.dir.fastq</code> folder, then run command:</p><pre><code class="language-Shell">ln -s file1
ln -s file2
...
</code></pre><p>or if you would like to link all fastq files in one folder, you can simply run:</p><pre><code class="language-Shell">ln -s [your folder path]/*_R1_001.fastq .

</code></pre><p>Then put your library csv file in the <code class="highlighter-rouge">analysis.par$out.dir.library</code> folder. The 1st column must be gRNA name and 2nd column must be sequence. An example library file would be like:</p> <div class="table-wrapper"><table> <thead> <tr> <th>id</th> <th style="text-align: center">seq</th> <th style="text-align: right">gene</th> </tr> </thead> <tbody> <tr> <td>TargetMouse.sg1</td> <td style="text-align: center">AAAAAGAAATGCTCTACCAG</td> <td style="text-align: right">Ypf1</td> </tr> <tr> <td>TargetMouse.sg2</td> <td style="text-align: center">AAAACACATACGTCTGTGAG</td> <td style="text-align: right">Cwc12</td> </tr> <tr> <td>TargetMouse.sg3</td> <td style="text-align: center">AAAACCGAGCACCATCAATG</td> <td style="text-align: right">Lck2</td> </tr> </tbody> </table></div> <p>Then go to <code class="highlighter-rouge">analysis.par$out.dir.library</code> folder and run following command to create fasta file for library.</p><pre><code class="language-Shell">awk FNR-1 your_library.csv | awk -F "," '{print "&gt;"$1"\n"$2}' &gt; your_library.fa" 
</code></pre><p>Then you could move forward to mapping and collecting raw counts step, which could be executed either on a high performance computing platform or your local machine.</p> <h3 id="step12-mapping-and-collecting-raw-counts"> <a href="#step12-mapping-and-collecting-raw-counts" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step1.2 Mapping and collecting raw counts </h3> <p>First of all, please ensure <code class="highlighter-rouge">blat</code> was installed in your working environment, since we will be utilizing <code class="highlighter-rouge">blat</code> for library sequence mapping. (You can use blat version <a href="https://anaconda.org/bioconda/blat">here on conda</a>)</p> <p><strong>NOTE: Please make sure your R is able to recognize blat. Please check whether blat has been added to your local bin</strong></p> <p>Next, we wil need to execute <code class="highlighter-rouge">ScreenBEAM.raw.count</code> function in ScreenBEAM2 R package. This function will help mapping your fastq files to library and collect raw counts by calling BLAT from R. This function will execute 1)get basic statististics, 2) create fasta from fastq and run blat, and 3) collect raw counts sequentially, so it may take a while for large sequencing data or whole genome libraries. Please try using more powerful computing platform if this is your case.</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">analysis.par</span><span class="o">&lt;-</span><span class="n">ScreenBEAM.raw.count</span><span class="p">(</span><span class="n">analysis.par</span><span class="p">)</span><span class="w">
</span><span class="c1"># save your analysis results</span><span class="w">
</span><span class="n">save</span><span class="p">(</span><span class="n">analysis.par</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="n">par.path</span><span class="p">,</span><span class="w"> </span><span class="n">par.name</span><span class="p">))</span><span class="w">

</span></code></pre></div></div> <p><strong>NOTE: Also, some fastq files are ultra big, you may need to chop them into small fastq files in order to execute them successfully.</strong></p> <h3 id="step13-mapping-qualtiy-control"> <a href="#step13-mapping-qualtiy-control" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step1.3 Mapping qualtiy control </h3> <p>After mapping is done, please proceed your analysis back in interactive environment (such as R studio). We are going to perform quality control, data annotation and data cleaning</p> <p>For mapping quality control, we designed a customized QC rmarkdown file, if you have ScreenBEAM2 installed, you can run function ‘ScreenBEAM.mapping.QC’ as follows:</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ScreenBEAM.mapping.QC</span><span class="p">(</span><span class="n">analysis.par</span><span class="p">)</span><span class="w">
</span></code></pre></div></div> <p>If you would like to make adjustment on this rmarkdown file, you can edit it and run your final rmarkdown file as follows:</p> <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ScreenBEAM.mapping.QC</span><span class="p">(</span><span class="n">analysis.par</span><span class="p">,</span><span class="w"> </span><span class="n">QC.Rmd.path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[</span><span class="n">new_path_to_your_rmd</span><span class="p">])</span><span class="w"> 
</span></code></pre></div></div> <p>This will output a QC report for your library mapping, here is <a href="./mapping_6sample_QC.html">example qc report</a>, including several useful metrics for mapping rate, mismatch rate, count boxplots, etc.</p> <h3 id="step14-read-your-mapping-quality-control"> <a href="#step14-read-your-mapping-quality-control" class="anchor-heading"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#link"></use></svg></a> Step1.4 Read your mapping quality control </h3> <p>Mapping quality control includes 6 differernt metrics to help you understand your library and mapping quality.</p> <p><strong>Library quality</strong></p> <ul> <li>Library removed RNA due to duplicated or substring of other RNAs This table shows all duplicated gRNAs(either perfect match or substring of other gRNA from the same gene). This will help you understand why some genes has less gRNAs than others.</li> </ul> <p><img src="./figures/table_duplicated_gRNA.png" alt="" /></p> <ul> <li>Total count of the gRNA number for each gene. This will help understand your library. Usually, a library contains hundreds or thousands of negative control gRNA (dummy gRNA) instead of 6(normally 6 gRNA for one gene). ScreenBEAM will take care of this unbalanced negative controls, and random sampling negative control guides to match other genes.</li> </ul> <p><img src="./figures/1.numofgRNAdistribution.png" alt="" /></p> <p><strong>Mapping quality</strong></p> <ul> <li>Mapping rate barplot. This barplot could help visualize how many reads in your samples’ fastq files are successfully mapped(perfect match) to library sequence. This could not only help you visualize the data quality of your fastq files, but also give a hint of whether to allow mismatch in your raw data collection.</li> </ul> <p><img src="./figures/2.mappingrate.png" alt="" /></p> <ul> <li>Boxplot of raw count distribution in each sample. This plot could help your understand if all samples have similar raw count distribution. It could also give you a hint of how to choose your total number to perform normalization.</li> </ul> <p><img src="./figures/rawCountDistribution.png" alt="" /></p> <ul> <li>Proportion of number of mismatch in each sample. This plot could help you understand if all samples have similar mapping quality.</li> </ul> <p><img src="./figures/nmismatch.png" alt="" /></p> <ul> <li>Suggestion of number of mismatch to choose for your downstream analysis. This plot visualizes the percentages of different number of mismatch in each samples respectively. If there are two suggested number of mismatch, please use the smaller one.</li> </ul> <p><img src="./figures/mismatchsuggestion.png" alt="" /></p> <hr> <footer role="contentinfo"> <p class="text-small text-grey-dk-000 mb-0">Copyright &copy; 2019- Yu Lab. Distributed by an <a href="https://github.com/jyyulab/scMINER/blob/master/LICENSE.txt">Apache License 2.0.</a></p> </footer> </div> </div> </div> </div> </body> </html>
