<h1 id="tutorial">Tutorial</h1>
<p>ScreenBEAM2 is a R based tool which consists of three major parts for processing steps: 1. mapping long read sequence to short read libraries, 2. Quality control, data cleanning and data preprocessing for mapped raw counts data; 3. Differential representative analysis on gene level or shRNA level.</p>

<h2 id="step1-mapping-to-library">Step1. Mapping to library</h2>
<p>First of all, define your project name and create a project R object. This could be achieved by function <code class="highlighter-rouge">ScreenBEAM.dir.create</code>.</p>

<p><strong>note: Project object should be library based! One library per run of ScreenBEAM2!</strong></p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">lib.name</span><span class="o">&lt;-</span><span class="s1">'Das_LibB_HSB2'</span><span class="w">
</span><span class="n">analysis.par</span><span class="o">&lt;-</span><span class="n">ScreenBEAM.dir.create</span><span class="p">(</span><span class="n">project_main_dir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'./'</span><span class="p">,</span><span class="w"> </span><span class="n">lib_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lib.name</span><span class="p">,</span><span class="w"> </span><span class="n">DATE</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">)</span><span class="w">

</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">par.path</span><span class="o">&lt;-</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">out.dir</span><span class="w">
</span><span class="n">analysis.par</span><span class="o">$</span><span class="n">par.name</span><span class="o">&lt;-</span><span class="n">paste0</span><span class="p">(</span><span class="s2">"analysis_par_"</span><span class="p">,</span><span class="n">lib.name</span><span class="p">,</span><span class="w"> </span><span class="s2">".RData"</span><span class="p">)</span><span class="w">

</span><span class="n">load</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="n">par.path</span><span class="p">,</span><span class="w"> </span><span class="n">par.name</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>
<p>Create a table with all raw fastq files that are mapped to this library</p>
<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">f</span><span class="o">&lt;-</span><span class="n">list.files</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">"../201901/rawdata/Das_fastq/Lib_B/"</span><span class="p">,</span><span class="n">patter</span><span class="o">=</span><span class="s2">".fastq$"</span><span class="p">,</span><span class="n">recursive</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="n">full.names</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">f</span><span class="o">&lt;-</span><span class="n">normalizePath</span><span class="p">(</span><span class="n">f</span><span class="p">)</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">f</span><span class="p">,</span><span class="n">file</span><span class="o">=</span><span class="n">file.path</span><span class="p">(</span><span class="n">par.path</span><span class="p">,</span><span class="s2">"fastq_paths.txt"</span><span class="p">),</span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">,</span><span class="n">quote</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="n">row.names</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="n">col.names</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w">

</span></code></pre></div></div>

<p>Next, soft link your fastq file to <code class="highlighter-rouge">analysis.par$out.dir.fastq</code> folder. This step is <strong>ESSENTIAL</strong>, please <strong>DON’T</strong> move files directly, in case of losing your data.</p>

<pre><code class="language-Shell">ln -s filepaths
</code></pre>

<p>Put library csv file with file name “JJ_Das_LibB.csv” in the <code class="highlighter-rouge">analysis.par$out.dir.library</code> folder. The 1st column must be gRNA name and 2nd column must be sequence. Then go to <code class="highlighter-rouge">analysis.par$out.dir.library</code> folder and run following command to create fasta file for library.</p>

<pre><code class="language-Shell">awk FNR-1 JJ_Das_LibB.csv | awk -F "," '{print "&gt;"$1"\n"$2}' &gt; JJ_Das_LibB.fa" 
</code></pre>

